### PR TITLE
Adding string calculator

### LINKED ISSUE #14: Exercise #7: String Calculator Kata
# String Calculator Kata

## Before you start
* Try not to read ahead.
* Do one scenario at a time. The trick is to learn to work incrementally.
* Make sure you only test for correct inputs. There is no need to test for invalid inputs for this kata.

### Step 1: the simplest thing
Create a simple String calculator with a method ``add(numbers)``.

* The string argument can contain 0, 1 or 2 numbers, and will return their sum (for an empty string it will return 0) for example ``""`` or ``"1"`` or ``"1,2"``.
* Start with the simplest test case of an empty string and move to 1 and two numbers.
* Remember to solve things as simply as possible so that you force yourself to write tests you did not think about.
* Remember to refactor after each passing test.

### Step 2: handle an unknown amount of numbers
Allow the ``add`` method to handle an unknown amount of numbers.

### Step 3: handle new lines between numbers
Allow the ``add`` method to handle new lines between numbers (instead of commas).

* the following input is ok:  ``"1\n2,3"`` (will equal 6)
* the following input is NOT ok:  ``"1,\n"`` (not need to prove it - just clarifying)

### Step 4: support different delimiters
Support different delimiters: to change a delimiter, the beginning of the string will contain a separate line that looks like this:   

``"//[delimiter]\n[numbers...]"``

For example ``"//;\n1;2"`` should return 3 where the default delimiter is ``';'``.

The first line is optional. 
All existing scenarios should still be supported.

### Step 5: negative numbers
Calling ``add`` with a negative number will throw an exception ``"negatives not allowed"`` - and the negative that was passed.

For example ``add("1,4,-1")`` should throw an exception with the message ``"negatives not allowed: -1"``.

If there are multiple negatives, show all of them in the exception message.

### Step 6: ignore big numbers
Numbers bigger than 1000 should be ignored, so adding 2 + 1001  = 2

---

Credits to [Roy Osherove](http://osherove.com/tdd-kata-1) for the original idea.
This is a Ruby adaptation from [here](https://github.com/xpeppers/string-calculator-kata/blob/master/README.md)

### PR DESCRIPTION
**Pull request by melissamohan**
Fixes #14

### CODE REVIEW

File: string_calculator.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
+class StringCalculator
+  def add(string_of_numbers = "")
+    sum = 0
+    separate_string = string_of_numbers.split(",")
+    numbers = separate_string.collect do |n|
+      n.to_i
+    end
+    neg = numbers.select(&:negative?)
+    is_negative = numbers.any? { |x| x.negative? }
```
Comment (snasirca) [inline string_calculator.rb]: If there is a negative number, you can just return early. You don't have to wait until line 22.

```ruby
return "negatives not allowed: #{neg.to_s}" if numbers.any?(&:negative?)
```

You can also use the shorthand by passing the symbol `:negative?` to `any?`.

File: string_calculator.rb [hunk 2]
```diff
@@ -0,0 +1,42 @@
+class StringCalculator
+  def add(string_of_numbers = "")
+    sum = 0
+    separate_string = string_of_numbers.split(",")
+    numbers = separate_string.collect do |n|
```
Comment (snasirca) [inline string_calculator.rb]: This block can be a single line block using curly braces to improve readability

File: string_calculator.rb [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+      n.to_i
+    end
+    neg = numbers.select(&:negative?)
+    is_negative = numbers.any? { |x| x.negative? }
+    is_greater_than_1000 = numbers.any? { |x| x >= 1000 }
```
Comment (snasirca) [inline string_calculator.rb]: Instead of doing this check, you can just build a list of `numbers` that has anything greater than 1000 excluded. So something like this would work too:

`numbers = numbers.reject { |number| number > 1000 }`. Also avoid using short variable names like `x` and `n` since they don't communicate what they are and hinder readability.

File: string_calculator.rb [hunk 4]
```diff
@@ -0,0 +1,42 @@
+class StringCalculator
+  def add(string_of_numbers = "")
+    sum = 0
+    separate_string = string_of_numbers.split(",")
```
Comment (snasirca) [inline string_calculator.rb]: I would postpone splitting the input string by comma until I've detected what the delimiter actually is. It may not be comma.

File: string_calculator.rb [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+      string_of_numbers_mod = string_of_numbers.gsub("//", "")
+      separator = string_of_numbers_mod[0].to_s
+      delimiters = ["\n", separator]
+      separator_split = string_of_numbers_mod.split(Regexp.union(delimiters))
+    else
+      delimiters = ["\n", ","]
+      comma_split = string_of_numbers.split(Regexp.union(delimiters))
+    end
+
+    if is_negative
```
Comment (snasirca) [inline string_calculator.rb]: If you handle these edge cases earlier, this set of if/else statements become much simpler.

File: string_calculator.rb [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    elsif is_greater_than_1000
+      less_than_1000.each { |x| sum += x }
+      sum
+    elsif string_of_numbers.length == 0
+      0
+    elsif string_of_numbers.length == 1
+      string_of_numbers[0].to_i
+    elsif string_of_numbers.length > 1
+      if string_of_numbers.start_with?("//")
+        separator_split.map!(&:to_i)
```
Comment (snasirca) [inline string_calculator.rb]: This far down in your algorithm, you shouldn't have to know that there is a "separator" split list of numbers or "comma" split list of numbers. You should just have a list of numbers that you can call `sum` on like this:

`numbers.sum` and return that.

File: spec/string_calculator_spec.rb [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+RSpec.describe StringCalculator do
+  it "returns 0 when empty" do
+    calculate = StringCalculator.new
+
+    result = calculate.add
+
+    expect(result).to eq(0)
+  end
+  it "returns number when one number is given" do
```
Comment (snasirca) [inline spec/string_calculator_spec.rb:12]: There should be some whitespace between `it` blocks.

File: spec/string_calculator_spec.rb [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+  end
+  it "returns the sum of two numbers" do
+    calculate = StringCalculator.new
+
+    result = calculate.add("4,8")
+
+    expect(result).to eq(12)
+  end
+  it "returns the sum of all numbers" do
+
```
Comment (snasirca) [inline spec/string_calculator_spec.rb:27]: Unnecessary whitespace here inside the `it` block.

**Review by snasirca — COMMENTED** [review]
This submission works but the algorithm is more complex than it needs to be. There should be phases in the algorithm that deal with individual concerns like this:

1. Determine what the delimiter is or use a default (comma) delimiter
2. Parse the rest of the input string using the delimiter into a list of `numbers` of type Integer
  2.1. Reject any `numbers` greater than 1000
3. Return early if a negative number is present in `numbers`
4. Return the sum of the `numbers`

In general, string parsing responsibilities should be dealt with early on and then not needed any more.

File: string_calculator.rb (hunk 1 above)
Comment (melissamohan) [reply string_calculator.rb]: Done, thank you for the tip

**Comment by melissamohan** [comment]
@snasirca , when you have a moment can you please review commit - refactoring after feedback session

Thank you
